tests(balancer) increase speed and reliability of balancer tests

hishamhm opened the pull request:
* Removes timed dependencies based on `ngx.sleep`: using the new `/health` endpoint, we can poll-wait on health changes and write tests that are timing-independent
* Avoids restarting Kong on every test: instead, this generates unique ids so that each test works on different entities.

Local results show a speed up from ~1min26s down to ~11s. :tada: 

The wrapper `describe` block is left unintented in the `spec-old-api` version to reduce `git diff` and to make the actual changes clearer (otherwise the whole file would be marked as changed). The one in `spec` is the one we will want to keep long-term and that one has been indented properly.

(Eventually we should remove the "postgres, cassandra" iteration from inside the tests anyway and just run one of each database in separate Travis jobs.)

thibaultcha commented on spec-old-api/02-integration/05-proxy/09-balancer_spec.lua:
```diff
@@ -235,234 +237,396 @@ local function http_server(timeout, host, port, counts, test_log)
 
 
-for ipv, localhost in pairs(localhosts) do
+dao_helpers.for_each_dao(function(kong_config)
```
minor but just as a reminder: while we still iterate over the DBs in the tests, and because this helper comes from `spec` (and not `old-spec`), then we should use the new iterator provided by `spec/helpers.lua` and meant to replace this callback approach. The iterator is easier to use and also more flexible (it allows one to easily run only the `postgres` or `cassandra` version of the tests by specifying an argument.

Since this is the old-spec suite that is bound to disappear, it may be ok to use it there for now...

thibaultcha commented on spec/02-integration/05-proxy/09-balancer_spec.lua:
```diff
@@ -371,194 +454,24 @@ local localhosts = {
 
 
-for mode, localhost in pairs(localhosts) do
+dao_helpers.for_each_dao(function(kong_config)
```
Here however, we really should rely on the proper utility. Combined with `get_db_utils()`, we get:
```lua
for _, strategy in helpers.each_strategy(strategy?) do
describe("<component> [#" .. strategy .. "]", function()

  setup(function()
    local bp, db, dao = helpers.get_db_utils(strategy) -- also runs migrations for us

    assert(helpers.start_kong({
      database   = strategy,
      nginx_conf = "spec/fixtures/custom_nginx.template",
    }))
  end)
end)
end
```

thibaultcha commented on spec-old-api/02-integration/05-proxy/09-balancer_spec.lua:
```diff
@@ -235,234 +237,396 @@ local function http_server(timeout, host, port, counts, test_log)
+end
+
+
+local function truncate_relevant_tables()
```
this could then receive an argument, `dao` and run the truncations from there instead of the very legacy `helpers.dao` instance

thibaultcha commented on spec/helpers.lua:
```diff
@@ -1163,8 +1173,13 @@ return {
     local ok, err = kong_exec("stop --prefix " .. prefix)
-    wait_pid(conf.nginx_pid, nil)
+
+    wait_pid(running_conf.nginx_pid)
```
This is a great fix. Maybe this warrants `--wait` flag in the CLI to allow for other users to not have to rely on custom logic when scripting `kong stop`? Maybe this should even be the default behaviour of `kong quit` (since it is the "graceful" stop).

hishamhm commented:
@thibaultcha Thank you for the review!

* reworded the helpers commit
* I added an optional `no_truncate` argument to `get_db_utils` because that truncates every table by default (speed difference was noticeable locally)
* changed to use `each_strategy` as you suggested, in both `spec` and `spec-old-api`
* got rid of `helpers.dao` in `truncate_relevant_tables`
* now that the code went through the initial round of reviews, I corrected the indentation of the `spec-old-api` version so that both files are in sync (now they are identical from `each_strategy` onwards)

> This is a great fix. Maybe this warrants --wait flag in the CLI to allow for other users to not have to rely on custom logic when scripting kong stop? Maybe this should even be the  default behaviour of kong quit (since it is the "graceful" stop).

To double-check if I understood this: you mean eventually moving `wait_pid` out of `helpers` (into `kong.tools.utils` or something) for use by `kong stop --wait` and `kong quit`?

thibaultcha commented:
Yay!

thibaultcha commented:
> To double-check if I understood this: you mean eventually moving wait_pid out of helpers (into kong.tools.utils or something) for use by kong stop --wait and kong quit?

Yes, this is what I meant (as a very low priority work item :sweat_smile:)
